toolchain/check/impl_lookup.cpp

```diff
@@ -347,6 +349,164 @@ static auto LookupImplWitnessInSelfFacetValue(
+// more specific value when possible.
+class SubstWitnessesCallbacks : public SubstInstCallbacks {
+ public:
+  // `context` must not be null.
+  explicit SubstWitnessesCallbacks(
+      Context* context, SemIR::LocId loc_id,
+      llvm::ArrayRef<SemIR::SpecificInterface> interfaces,
+      llvm::ArrayRef<SemIR::InstId> witness_inst_ids)
+      : SubstInstCallbacks(context),
+        loc_id_(loc_id),
+        interfaces_(interfaces),
+        witness_inst_ids_(witness_inst_ids) {}
+
+  auto Subst(SemIR::InstId& inst_id) -> SubstResult override {
+    // `FacetType` can be concrete even when it has rewrite constraints that
+    // have a symbolic dependency on `.Self`. See use of
+    // `GetConstantValueIgnoringPeriodSelf` in eval. So in order to recurse into
+    // `FacetType` we must check for it before the `is_concrete` early return.
+    if (context().insts().Is<SemIR::FacetType>(inst_id)) {
+      ++facet_type_depth_;
+      return SubstOperands;
+    }
+
+    if (context().constant_values().Get(inst_id).is_concrete()) {
+      return FullySubstituted;
+    }
+
+    auto access = context().insts().TryGetAs<SemIR::ImplWitnessAccess>(inst_id);
+    if (!access) {
+      return SubstOperands;
+    }
+
+    auto lookup =
+        context().insts().GetAs<SemIR::LookupImplWitness>(access->witness_id);
+    auto bind_name = context().insts().TryGetAs<SemIR::BindSymbolicName>(
+        lookup.query_self_inst_id);
+    if (!bind_name) {
+      return SubstOperands;
+    }
+
+    const auto& self_entity_name =
+        context().entity_names().Get(bind_name->entity_name_id);
+    if (self_entity_name.name_id != SemIR::NameId::PeriodSelf) {
+      return SubstOperands;
+    }
+
+    // TODO: Once we are numbering `EntityName`, (see the third model in
+    // https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.7urbxcq23olv)
+    // then verify that the index here is equal to the `facet_type_depth_`,
+    // which would mean that it is a reference to the top-level `Self`, which is
+    // being replaced with the impl lookup query self facet value (and then we
+    // use the witness derived from it).
+    //
+    // For now, we only substitute if depth == 0, which is incorrect inside
+    // nested facet types, as it can miss references in specifics up to the top
+    // level facet value.
+    if (facet_type_depth_ > 0) {
+      return SubstOperands;
+    }
+
+    auto witness_id =
+        FindWitnessForInterface(lookup.query_specific_interface_id);
+    if (!witness_id.has_value()) {
+      return SubstOperands;
+    }
+
+    inst_id = RebuildNewInst(
+        context().insts().GetLocIdForDesugaring(loc_id_),
+        SemIR::ImplWitnessAccess{.type_id = GetSingletonType(
+                                     context(), SemIR::WitnessType::TypeInstId),
+                                 .witness_id = witness_id,
+                                 .index = access->index});
```

Comment:
Looking at this, I wonder if we could match a `LookupImplWitness` and replace it with the result from `FindWitnessForInterface` instead of matching an `ImplWitnessAccess` and replacing it with an `ImplWitnessAccess` with a different witness. Or do we only want to do the rewrite when the witness is consumed by an `ImplWitnessAccess` and not if it's used in some other context (eg, forming a facet value)?